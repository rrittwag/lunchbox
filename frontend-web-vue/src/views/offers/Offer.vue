<script setup lang="ts">
import type { LunchOffer } from '@/model/lunch'
import { formatEuro } from '@/util/formatting'
import Badge from '@/views/offers/Badge.vue'
import { computed } from 'vue'

const props = defineProps<{
  offer: LunchOffer
  showDetailsInXS?: boolean
}>()

const sortedTags = computed(() => props.offer.tags.slice().sort())
const priceAsString = computed(() => formatEuro(props.offer.price))
const isVeggie = (tagLabel: string) => tagLabel === 'vegetarisch' || tagLabel === 'vegan'
</script>

<template>
  <li class="w-full pt-3 sm:pt-4">
    <div class="flex w-full items-baseline">
      <h4 class="grow text-xl leading-tight text-neutral-800">
        {{ props.offer.name }}
      </h4>
      <span v-if="props.offer.price" aria-label="Preis" class="shrink-0 pl-2 text-xl leading-tight text-neutral-800">
        <small class="pl-2 pr-1 text-lg font-light text-neutral-800">â‚¬</small>{{ priceAsString }}
      </span>
    </div>

    <p
      role="note"
      class="pl-px pt-px font-light leading-snug text-neutral-600 sm:block"
      :class="props.showDetailsInXS ? 'flex' : 'hidden'"
    >
      {{ props.offer.description }}
    </p>

    <p class="pt-1 sm:flex" :class="props.showDetailsInXS ? 'flex' : 'hidden'">
      <Badge
        v-for="tagLabel in sortedTags"
        :key="tagLabel"
        role="note"
        :label="tagLabel"
        :color="isVeggie(tagLabel) ? 'bg-success-200' : 'bg-accent-200'"
      />
    </p>
  </li>
</template>
