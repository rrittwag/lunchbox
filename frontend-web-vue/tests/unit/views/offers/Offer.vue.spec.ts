import Offer from '@/views/offers/Offer.vue'
import { gyros } from '@tests/unit/test-data'
import { render } from '@testing-library/vue'

describe('Offer', () => {
  it('renders', () => {
    const { getAllByRole, getByRole, getByLabelText } = render(Offer, {
      props: { offer: gyros },
    })

    expect(getByRole('heading', { level: 4 })).toHaveTextContent(gyros.name)
    expect(getByLabelText('Preis')).toHaveTextContent('â‚¬3,50')
    const notes = getAllByRole('note')
    expect(notes).toHaveLength(3)
    expect(notes[0]).toHaveTextContent(gyros.description)
    expect(notes[1]).toHaveTextContent(gyros.tags[0])
    expect(notes[2]).toHaveTextContent(gyros.tags[1])
  })

  it('renders offer without price', () => {
    const gyrosWithoutPrice = { ...gyros, price: undefined }
    const { getAllByRole, getByRole, queryByLabelText } = render(Offer, {
      props: { offer: gyrosWithoutPrice },
    })

    expect(getByRole('heading', { level: 4 })).toHaveTextContent(gyros.name)
    expect(queryByLabelText('Preis')).not.toBeInTheDocument()
    const notes = getAllByRole('note')
    expect(notes).toHaveLength(3)
    expect(notes[0]).toHaveTextContent(gyros.description)
    expect(notes[1]).toHaveTextContent(gyros.tags[0])
    expect(notes[2]).toHaveTextContent(gyros.tags[1])
  })

  it('hides details for screen size XS', () => {
    const { getByRole, getByLabelText } = render(Offer, {
      props: { offer: gyros },
    })

    expect(getByRole('heading', { level: 4 })).toBeVisible()
    expect(getByLabelText('Preis')).toBeVisible()
    // FIXME: Tailwind classes do not work with testing-library -> https://stackoverflow.com/a/74160802
    // getAllByRole('note').forEach((elem) => expect(elem).not.toBeVisible())
  })

  it('shows details for screen size XS', () => {
    const { getAllByRole, getByRole, getByLabelText } = render(Offer, {
      props: { offer: gyros, showDetailsInXS: true },
    })

    expect(getByRole('heading', { level: 4 })).toBeVisible()
    expect(getByLabelText('Preis')).toBeVisible()
    getAllByRole('note').forEach((elem) => expect(elem).toBeVisible())
  })
})
