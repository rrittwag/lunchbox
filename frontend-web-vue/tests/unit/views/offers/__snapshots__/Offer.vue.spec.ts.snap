// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offer renders snapshot 1`] = `
<li
  class="w-full pt-3 sm:pt-4"
>
  <div
    class="flex w-full items-baseline"
  >
    <h4
      class="grow text-xl leading-tight text-neutral-800"
    >
      Gyros
    </h4>
    <span
      class="shrink-0 pl-2 text-xl leading-tight text-neutral-800"
    >
      <small
        class="pl-2 text-lg font-light text-neutral-800"
      >
         â‚¬ 
      </small>
       3,50
    </span>
  </div>
  <span
    class="pt-px pl-px font-light leading-snug text-neutral-600 sm:block hidden"
  >
    mit Pommes
  </span>
  <div
    class="pt-1 sm:flex hidden"
  >
    
    <badge-stub
      color="bg-accent-200"
      label="Tagesgericht"
    />
    <badge-stub
      color="bg-accent-200"
      label="lecker"
    />
    
  </div>
</li>
`;
